tuning/utils/merge_model_utils.py

```diff
@@ -102,3 +102,77 @@ def fetch_base_model_from_checkpoint(checkpoint_model: str) -> str:
             "Base model adapter config exists, but has no base_model_name_or_path!"
         )
     return adapter_dict["base_model_name_or_path"]
+
+def _copy_files_to_directory(src: str , dest: str, exclude_files:list[str] = []):
+    import shutil
+    src_files = os.listdir(src)
+    for file_name in src_files:
+        if file_name in exclude_files:
+            continue
+        full_file_name = os.path.join(src, file_name)
+        if os.path.isfile(full_file_name):
+            shutil.copy(full_file_name, dest)
+
+def post_process_vLLM_adapters_new_tokens(path_to_checkpoint: str, modified_checkpoint_path:str = None):
+
+    from safetensors import safe_open
+    from safetensors.torch import save_file
+
+    # if not set, original checkpoint will be modified
+    if not modified_checkpoint_path:
+        modified_checkpoint_path = path_to_checkpoint
+
+    # Get all values of new token indexes
+    sorted_token_indexes = []
+    if os.path.isfile(os.path.join(path_to_checkpoint,"added_tokens.json")):
+        with open(os.path.join(path_to_checkpoint,"added_tokens.json"), "r") as fp:
+            added_tokens = json.load(fp)
+            sorted_token_indexes = sorted(added_tokens.values())
+
+    with safe_open(os.path.join(path_to_checkpoint, "adapter_model.safetensors"), framework="pt") as f:
+        new_embeddings = {}
+        adapters = {}
+        embeddings_weights_in_adapters = False
+        # Quickly check if post-processing is needed by checking adapters file for weights
+        for k in f.keys():
+            if 'lm_head.weight' in k or 'embed_tokens.weight' in k:
+                embeddings_weights_in_adapters = True
+                if len(sorted_token_indexes) >=1:
+                    raise NotImplementedError("Seems like embeddings are resized without adding new tokens. \
+                                               Cannot be post-processed to load on vLLM. Try setting \
+                                               parameter `embedding_size_multiple_of` to 1" )
+
+        # Post-processing is needed to copy out new vectors
+        if embeddings_weights_in_adapters:
+            for k in f.keys():
+                if 'lm_head.weight' in k:
```

Comment:
`lm_head` is not a standard name. It is the name used in most models, but there might be differences say from LLama and GPTBIgCode (which the older granite models depend on)

Reply:
maybe a suggestion

```python
def infer_head_embedding_names_from_vocab_size(vocab_size):
    try:
        # get (names, shapes) of all the modules that have one dimension that matches the vocab size
        name_and_sizes = [
            (name.split('.'), tuple(mod.weight.shape)) for name, mod in model.named_modules()
            if hasattr(mod, 'weight') and vocab_size in tuple(mod.weight.shape)
        ]

        # if there are more than 2 items left, then all bets are off
        if len(name_and_sizes) != 2:
            raise IndexError

        # sort them by the number of levels
        name_and_sizes = sorted(sizes, key= lambda x: len(x[0]))

        # the heuristic is that the lm_head will usually be at the CausalLM, so it will be at 
        # the lowest level, and the embedding will be inside the LM and at a higher level
        lm_head_name, _ = name_and_sizes[0]
        embed_name, _  = name_and_sizes[1]
        return lm_head_name[-1], embed_name[-1]
    except IndexError:
        return 'lm_head', 'embed_tokens' 

# usage, either use model.config.vocab_size, or if you know the resized vocab, use that
infer_head_embedding_names_from_vocab_size(model.config.vocab_size)
```

Reply:
thats a good idea, thank you!

Reply:
thanks for suggestion @fabianlim . We tested with every unique model architecture we support as @willmj documented https://github.com/foundation-model-stack/fms-hf-tuning/pull/338#issuecomment-2364505870 and all the models had these standard names. 

Hence we decided that inferring names is lower priority and we will add it an enhancement in future PR after this release - this PR is currently blocking release